mrc-2380 Allow editing of standard form parameters

EmmaLRussell opened the pull request:
- This branch enables editing of dynamic form parameters via modal dialogs. When the user OKs from dialog, results are re-fetched and the view updated. An 'Updating analysis..' indicator is displayed while results are being fetched, both on initial fetch and after parameter changes. Try updating the healthcare capacity values, and see the horizontal lines update on the Healthcare demand chart. 
- The 'Edit' buttons appear unattractively bolted below each collapsed readonly parameter form, rather than being rolled into the collapsed section. I'm going to address this in a separate branch by implementing the collapsible panels in the app rather than using the collapse functionality in the dynamic form (we'll need this for the phases parameters anyway). 
- I'm following the pattern already implicitly started where the Home component interacts with Vuex store while its child components just accept props and emit events. I feel like this makes the components easier to test, and makes the Vuex interactions easier to locate, but it's more long-winded (Parameters component could directly update the store rather than going via Home.)
- Not yet implemented is extracting the default parameter values from parameter metadata - initial values are still hard-coded until a future branch. 
- I've used a couple of handy components copied direct from HINT - LoadingSpinner and Modal. 
- I'm not entirely sure why the barchart seems to stay behind while the other charts refresh - possibly because that's the only chart which has not changed. That suggests there might be some performance issues in the chart rendering.

EmmaLRussell committed:
Edit parameters dialog, and refetch results

EmmaLRussell committed:
Add fetching indicator

EmmaLRussell committed:
Add fetching indicator

EmmaLRussell committed:
Merge branch 'main' into mrc-2380

EmmaLRussell committed:
Add editParameters test

EmmaLRussell committed:
Added tests

EmmaLRussell committed:
linting

EmmaLRussell committed:
Add spinner tests

EmmaLRussell committed:
lint

EmmaLRussell committed:
reinstate background color

mwoodbri reviewed (changes requested):
Very exciting to see a fully-working app (well, at least a version 0.1).

I haven't gone through code line-by-line because wanted to document some behaviour that probably needs resolving first:
* Selecting "No Prioritisation" for Prioritisation & Allocation causes an error - though I appreciate this might be due to cometr being WIP?
* "Total general beds" (for example) says it's required but if you leave it empty and click submit you end up in a bad place
* Charts: I think you're right - the bottom bar chart isn't changing so isn't re-rendered (I guess this is expected, given the state of cometr?). But might be generally best to display either the spinner or the chart panel to avoid this behaviour.
* Trying to submit an empty value for required % field isn't allowed, but produces a big error console message regardless, which I think is because it's trying to divide an empty number by 100.
* "Vaccine Durability" doesn't say "required" but throws an error if omitted
* More generally there is a conspicuous lack of error handling when POSTing parameters to model

Some more cosmetic points - maybe best addressed by another ticket if even relevant:
* Still some wasted space in tables - would be best if left-hand column expanded as far as possible without causing right-hand column to wrap. This would give a better chance of long labels not being wrapped.
* Big numbers would be much more readable if displayed with separators, at least in read-only view
* I found myself reaching for a "Restore defaults" button - might be nice if possible either in dialog or at top-level - for when doing exploratory analysis. Can just reload the page, but I guess we might implement persistence at some point, meaning this wouldn't work.
* I also found myself expanding the forms every time they were closed, to cross-reference what I was seeing in the charts with the form values. But I guess keeping them all open wouldn't make so much sense when there are too many parameter values to fit on one screen!

Nicolas-Dolan commented on src/app/static/src/components/parameters/EditParameters.vue:
```diff
@@ -0,0 +1,91 @@
+        Modal,
+        DynamicForm
+    },
+    setup(props: Props, context) {
```
I've not seen this syntax before.

EmmaLRussell replied:
Welcome to [the future](https://v3.vuejs.org/guide/composition-api-introduction.html) :)

EmmaLRussell commented:
>   Selecting "No Prioritisation" for Prioritisation & Allocation causes an error - though I appreciate this might be due to cometr being WIP?
> 
I've changed this to 'All' which is a value [supported by nimue](https://github.com/mrc-ide/nimue/blob/b38aa37f279528a1c45c3a2b20805857d11b0de0/R/prioritisation_matrix.R) but there are other discrepancies - not sure why HCW and Elderly is working, but HCW, Elderly and High-Risk Groups fails. I'm going to make a list of queries about parameters I'm still not sure about and try to get some answers on those later this week. 

>  "Total general beds" (for example) says it's required but if you leave it empty and click submit you end up in a bad place

I completely forgot to deal with invalid forms - now fixed by disabling the OK button if the form is invalid (basically if a required value is missing). 

However, I think we have some further work to do in the form - while it won't accept a 0 value for a required value (I guess because it's falsy - though that might be a valid numerical value!) - it will quite happily validate with a value less than the min or more than the max specified. I think the min and max are just used to configure the inputs. This means that we can avoid having a divide-by-zero bad time, but we continue to have bad times caused by having a vaccine which is 200% effective or -5% effective..  

I think the form should be configurable to not validate if bounds exceeded, and in this case show show a message indicating why. I've made a ticket for this: https://mrc-ide.myjetbrains.com/youtrack/issue/RESIDE-236

> 
>   Charts: I think you're right - the bottom bar chart isn't changing so isn't re-rendered (I guess this is expected, given the state of cometr?). But might be generally best to display either the spinner or the chart panel to avoid this behaviour.
> 

OK, now hiding the charts entirely while updating. 

>   Trying to submit an empty value for required % field isn't allowed, but produces a big error console message regardless, which I think is because it's trying to divide an empty number by 100.

See remark above - this problem is now avoided, but other values are an issue. 

> 
>   "Vaccine Durability" doesn't say "required" but throws an error if omitted

This value now required. 

>   More generally there is a conspicuous lack of error handling when POSTing parameters to model

Indeed there's a conspicuous lack of error handling in all requests so far! Ticket to address this: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2409

> Some more cosmetic points - maybe best addressed by another ticket if even relevant:
> 
>   Still some wasted space in tables - would be best if left-hand column expanded as far as possible without causing right-hand column to wrap. This would give a better chance of long labels not being wrapped.

This is something of a limitation of the dynamic form as it puts bootstrap `col-n` classes on elements, which are a pain to override. There is a ticket to use semantic classes instead which will make tweaking this behaviour much easier: https://mrc-ide.myjetbrains.com/youtrack/issue/RESIDE-212

> 
>   Big numbers would be much more readable if displayed with separators, at least in read-only view

Agreed - again, I think that needs to be fixed in the form. https://mrc-ide.myjetbrains.com/youtrack/issue/RESIDE-237

> 
>   I found myself reaching for a "Restore defaults" button - might be nice if possible either in dialog or at top-level - for when doing exploratory analysis. Can just reload the page, but I guess we might implement persistence at some point, meaning this wouldn't work.
> 

This is in the mockup, so we'll get to it at some point, but it's a good point, as we should retain the default parameter values so we can reinstate them later if required. 

>   I also found myself expanding the forms every time they were closed, to cross-reference what I was seeing in the charts with the form values. But I guess keeping them all open wouldn't make so much sense when there are too many parameter values to fit on one screen!

Well that's the idea... I guess we'll get user feedback if the preference would be for everything to be always visible after all.

Nicolas-Dolan reviewed (commented):
This is a good first start. I had to delete my local comet repo and start again to get it to work; probably something to do with the vue 3 to vue 2 migration. I noticed a couple of things that don't need to be addressed in this ticket such as a lack of error page and the favicon needing updating. Also would be good to be able to cancel an updating analysis if it gets stuck as it did once while I was testing it.

EmmaLRussell committed:
Address PR comments

EmmaLRussell commented:
> This is a good first start. I had to delete my local comet repo and start again to get it to work; probably something to do with the vue 3 to vue 2 migration. I noticed a couple of things that don't need to be addressed in this ticket such as a lack of error page and the favicon needing updating. Also would be good to be able to cancel an updating analysis if it gets stuck as it did once while I was testing it.

Thanks Nick - you'd definitely need to re-do npm install, having to delete the whole repo sounds a bit drastic. 

Error handling to be added in this ticket, which should fix the 'analysis getting stuck' problem: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2409

When you say no error page, do you mean like a 404 page, or just no errors displayed in-app?

Yep, favicon is on the todo list...

mwoodbri reviewed (approved):
LGTM.

We may ultimately have to set a min width for the chart legends to ensure that the phases line up on each chart, and maybe set an initial width for the parameters pane so that the charts don't skip when first loading the page. But this is much further down the line!

Nicolas-Dolan commented:
> When you say no error page, do you mean like a 404 page, or just no errors displayed in-app?

The 404 page. When I used an unrecognised url I got a generic error page rather than one generated by the app.

EmmaLRussell commented:
> > When you say no error page, do you mean like a 404 page, or just no errors displayed in-app?
> 
> The 404 page. When I used an unrecognised url I got a generic error page rather than one generated by the app.

Ticket added: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-2417

Nicolas-Dolan commented on src/app/static/src/assets/custom.scss:
```diff
@@ -32,4 +32,26 @@ $body-bg: #eee;
+  @include button-variant($theme-blue, darken($theme-blue, 10%))
+}
+
+#fetching-results {
+  background-color: rgba(0,0,0,0.2);
+  position: absolute;
+  left: 0;
+  top: 0;
+  width: 100%;
+  height: 100vh;
+}
+
+#fetching-results-msg {
+  margin-left: auto;
+  margin-right: auto;
+  margin-top: 5rem;
+  width: 14rem;
+  background: white;
+  padding: 1rem;
+}
```
If these css styling are only pointing at one component, it might be worth considering putting them in a scoped style tag inside the component itself rather than a global style sheet. I understand we haven't been doing this so far but I think it might be a little easier to follow.

EmmaLRussell replied:
Interesting - my tendency is to keep css all together separate from the markup, but if consensus was to scope to components where possible we could try that instead.
